v5.1

jbender opened the pull request:
Merges in a bunch of upstream fixes, and starts to get specs running again

jbender committed:
bind_visitor was removed in Arel v9

jbender committed:
enable customization of dsn connections

jbender committed:
add snowflake adapter

jbender committed:
leverage build matrix

jbender committed:
add connections test

https://github.com/localytics/odbc_adapter/blob/622105e07c7d8a80f0940daa93d78d15bc315180/test/connections_test.rb

jbender committed:
reverse quotes

jbender committed:
errors only take one argument

jbender committed:
complete removal of Arel::Visitors::BindVisitor

jbender committed:
add connection string test

https://github.com/localytics/odbc_adapter/blob/a30f0c09e2613a9936155bf50be82ceaf6719ea0/test/connection_string_test.rb

jbender committed:
fix rubocop errors

jbender committed:
revert use of tap

jbender committed:
update assertions to reflect attr mass assignment

jbender committed:
simplify expectation of assignment

jbender committed:
add gemfiles to test multiple activerecord versions

jbender committed:
fix travis bundler

jbender committed:
fix gemspec path

jbender committed:
force bundler downgrade

jbender committed:
more specific rake tasks

jbender committed:
move required gems into gemspec

jbender committed:
disable prepared statements in postgres

jbender committed:
add extension docs

jbender committed:
Travis fixups

jbender commented:
@rnaud / @nickelser This should be good to go, and for the first time, with passing specs to prove it.

https://travis-ci.org/jbender/odbc_adapter/builds/493921681

BTW, I don't have sufficient permissions to push directly/enable the test suite on Travis for this repo, but we should probably do that then tie in passing builds to PRs

nickelser commented on lib/odbc_adapter/adapters/mysql_odbc_adapter.rb:
```diff
@@ -6,7 +6,7 @@ class MySQLODBCAdapter < ActiveRecord::ConnectionAdapters::ODBCAdapter
 
       class BindSubstitution < Arel::Visitors::MySQL
-        include Arel::Visitors::BindVisitor
+        include Arel::Visitors::BindVisitor if Arel::VERSION.to_i < 9
```
might be good to have a comment on this restriction

nickelser commented on .travis.yml:
```diff
@@ -1,29 +1,43 @@
+  - sh -c "if [ '$DB' = 'mysql' ]; then mysql -e 'GRANT ALL PRIVILEGES ON *.* TO "root"@"localhost";' -uroot; fi"
+
+  - sh -c "if [ '$DB' = 'postgresql' ]; then sudo odbcinst -i -d -f /usr/share/psqlodbc/odbcinst.ini.template; fi"
+  - sh -c "if [ '$DB' = 'postgresql' ]; then psql -c 'CREATE DATABASE odbc_test;' -U postgres; fi"
```
ideally we would have some test account w/snowflake but i see that being somewhat difficult (and definitely not in scope here :)

nickelser commented on lib/odbc_adapter/database_statements.rb:
```diff
@@ -128,7 +128,7 @@ def nullability(col_name, is_nullable, nullable)
 
     def prepared_binds(binds)
-      prepare_binds_for_database(binds).map { |bind| _type_cast(bind) }
+      binds.map(&:value_for_database).map { |bind| _type_cast(bind) }
```
i'm not following this change here -- what's wrong with the `prepare_binds_for_database_method`?

jbender replied:
I don't remember exactly what prompted it, but it's an internal method that's been removed from Rails:

https://github.com/rails/rails/commit/fb30df9c10372b1667dabac2fffe2cff25302626

nickelser reviewed (approved):
overall this looks great, with some changes that don't fully make sense to me (but don't look wrong either 😄)

jbender commented:
@nickelser I cleaned up the history and split this hydra into their respective parts (adding a comment about the Arel issue). If you want to go ahead and merge those, I'd recommend doing them in the order that I added the PRs.

nickelser commented:
sounds good, thanks for doing this! will merge now
